presto-main/src/main/java/io/prestosql/operator/aggregation/builder/SpillableHashAggregationBuilder.java

```diff
@@ -173,6 +188,21 @@ private boolean shouldMergeWithMemory(long memorySize)
     public WorkProcessor<Page> buildResult()
     {
         checkState(hasPreviousSpillCompletedSuccessfully(), "Previous spill hasn't yet finished");
+        producingOutput = true;
+
+        // convert revocable memory to user memory as memory must not be revoked when aggregation operator is finishing
+        if (localRevocableMemoryContext.getBytes() > 0) {
+            long currentRevocableBytes = localRevocableMemoryContext.getBytes();
+            localRevocableMemoryContext.setBytes(0);
+            if (!localUserMemoryContext.trySetBytes(hashAggregationBuilder.getSizeInMemory())) {
+                // TODO: this might fail (even though we have just released memory), but we don't
+                // have a proper way to atomically convert memory reservations
+                localRevocableMemoryContext.setBytes(currentRevocableBytes);
+                // spill since revocable memory could not be converted to user memory immediately
+                getFutureValue(spillToDisk());
```

Comment:
This can take time.
Can we spill asynchronously?
```
spillToDisk();
return WorkProcessor.empty();
```
(then we need to `updateMemory();` on next call..)

Reply:
Added todo since this requires bigger operator refactor